### PR TITLE
inlist: move type coercion to logical phase

### LINKED ISSUE #3468: move type coercion for inlist from physical phase to logical phase
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

when do https://github.com/apache/arrow-datafusion/pull/3396, the `prune_decimal_in_list` test case can't be passed.

We should do the type coercion for inlist in the logical phase.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #3468

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/planner.rs [hunk 1]
```diff
@@ -410,10 +408,7 @@ pub fn create_physical_expr(
                         )
                     })
                     .collect::<Result<Vec<_>>>()?;
-
-                let (cast_expr, cast_list_exprs) =
-                    in_list_cast(value_expr, list_exprs, input_schema)?;
-                expressions::in_list(cast_expr, cast_list_exprs, negated, input_schema)
+                expressions::in_list(value_expr, list_exprs, negated, input_schema)
```
Comment (liukun4515) [inline datafusion/physical-expr/src/planner.rs:411]: type coercion done in the logical phase

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -960,6 +960,17 @@ pub fn in_list(
     negated: &bool,
     schema: &Schema,
 ) -> Result<Arc<dyn PhysicalExpr>> {
+    // check the data type
```
Comment (liukun4515) [inline datafusion/physical-expr/src/expressions/in_list.rs:963]: check the data type and make sure the same type for value expr and list exprs.

File: datafusion/optimizer/src/type_coercion.rs [hunk 3]
```diff
@@ -164,11 +163,61 @@ impl ExprRewriter for TypeCoercionRewriter<'_> {
...
+                expr,
+                list,
+                negated,
+            } => {
+                let expr_data_type = expr.get_type(&self.schema)?;
+                let list_data_types = list
+                    .iter()
+                    .map(|list_expr| list_expr.get_type(&self.schema))
+                    .collect::<Result<Vec<_>>>()?;
+                let result_type =
```
Comment (liukun4515) [inline datafusion/optimizer/src/type_coercion.rs:176]: Do we need a common function in a new file like `in_list_rule.rs` to do the `coerce_types`?
cc @alamb @andygrove

Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:176]: it is also can be used in the `in_list.rs` for physical inlist test cases.

File: datafusion/core/src/physical_plan/planner.rs [hunk 4]
```diff
@@ -1694,9 +1694,11 @@ mod tests {
     async fn plan(logical_plan: &LogicalPlan) -> Result<Arc<dyn ExecutionPlan>> {
         let mut session_state = make_session_state();
         session_state.config.target_partitions = 4;
+        // optimize the logical plan
+        let logical_plan = session_state.optimize(logical_plan)?;
```
Comment (liukun4515) [inline datafusion/core/src/physical_plan/planner.rs:1698]: after the optimization, the type coercion could be done.

File: datafusion/core/src/physical_plan/planner.rs [hunk 5]
```diff
@@ -1994,7 +1994,10 @@ mod tests {
...
+        let expected = "expr: [(InListExpr { expr: Column { name: \"c1\", index: 0 }, list: [Literal { value: Utf8(\"a\") }, Literal { value: Utf8(\"1\") }, Literal { value: Utf8(\"2\") },";
```
Comment (andygrove) [inline datafusion/core/src/physical_plan/planner.rs:1997]: Well that is much easier to read :heart:

**Review by Dandandan — APPROVED** [review]

**Review by alamb — COMMENTED** [review]
This looks great @liukun4515

File: datafusion/core/src/physical_plan/planner.rs (hunk 5 above)
Comment (liukun4515) [reply datafusion/core/src/physical_plan/planner.rs:1997]: yes, i only show the prefix and suffix of the output

File: datafusion/core/src/physical_plan/planner.rs [hunk 6]
```diff
@@ -1712,12 +1714,12 @@ mod tests {
...
             .build()?;
 
-        let plan = plan(&logical_plan).await?;
+        let exec_plan = plan(&logical_plan).await?;
 
         // verify that the plan correctly casts u8 to i64
         // the cast here is implicit so has CastOptions with safe=true
-        let expected = "BinaryExpr { left: Column { name: \"c7\", index: 6 }, op: Lt, right: TryCastExpr { expr: Literal { value: UInt8(5) }, cast_type: Int64 } }";
-        assert!(format!("{:?}", plan).contains(expected));
+        let expected = "BinaryExpr { left: Column { name: \"c7\", index: 2 }, op: Lt, right: Literal { value: Int64(5) } }";
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply datafusion/core/src/physical_plan/planner.rs:1721]: that optimization was done by @Dandandan in https://github.com/apache/arrow-datafusion/pull/3444

File: datafusion/optimizer/src/type_coercion.rs (hunk 3 above)
Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:176]: I will a issue to track this and centralize the code together for the inlist

**Pull request merged**